Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp

```diff
@@ -0,0 +1,146 @@
+#include "WebDriverBidiProtocolObjects.h"
+#include <JavaScriptCore/InspectorBackendDispatcher.h>
+#include <JavaScriptCore/InspectorFrontendRouter.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/WeakPtr.h>
+
+namespace WebKit {
+
+using namespace Inspector;
+using BrowsingContext = Inspector::Protocol::BidiBrowsingContext::BrowsingContext;
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(WebDriverBidiProcessor);
+
+WebDriverBidiProcessor::WebDriverBidiProcessor(WebAutomationSession& session)
+    : m_session(session)
+    , m_frontendRouter(FrontendRouter::create())
+    , m_backendDispatcher(BackendDispatcher::create(m_frontendRouter.copyRef()))
+    , m_browserDomainDispatcher(BidiBrowserBackendDispatcher::create(m_backendDispatcher, this))
+    , m_browsingContextDomainDispatcher(BidiBrowsingContextBackendDispatcher::create(m_backendDispatcher, this))
+    , m_logDomainNotifier(makeUnique<BidiLogFrontendDispatcher>(m_frontendRouter))
+{
+    protectedFrontendRouter()->connectFrontend(*this);
+}
+
+WebDriverBidiProcessor::~WebDriverBidiProcessor()
+{
+    protectedFrontendRouter()->disconnectFrontend(*this);
+}
+
+Ref<Inspector::FrontendRouter> WebDriverBidiProcessor::protectedFrontendRouter() const
+{
+    return m_frontendRouter;
+}
+
+Ref<Inspector::BackendDispatcher> WebDriverBidiProcessor::protectedBackendDispatcher() const
+{
+    return m_backendDispatcher;
+}
+
+void WebDriverBidiProcessor::processBidiMessage(const String& message)
+{
+    RefPtr session = m_session.get();
+    if (!session) {
+        LOG(Automation, "processBidiMessage of length %d not delivered, session is gone!", message.length());
+        return;
+    }
+
+    LOG(Automation, "[s:%s] processBidiMessage of length %d", session->sessionIdentifier().utf8().data(), message.length());
+    LOG(Automation, "%s", message.utf8().data());
+
+    protectedBackendDispatcher()->dispatch(message);
+}
+
+void WebDriverBidiProcessor::sendBidiMessage(const String& message)
+{
+    RefPtr session = m_session.get();
+    if (!session) {
+        LOG(Automation, "sendBidiMessage of length %d not delivered, session is gone!", message.length());
+        return;
+    }
+
+    LOG(Automation, "[s:%s] sendBidiMessage of length %d", session->sessionIdentifier().utf8().data(), message.length());
+    LOG(Automation, "%s", message.utf8().data());
+
+    session->sendBidiMessage(message);
+}
+
+
+// MARK: Inspector::FrontendChannel methods.
+
+void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+{
+    sendBidiMessage(message);
+}
+
+
+// MARK: Inspector::BrowsingContextDispatcherHandler methods.
+
+void WebDriverBidiProcessor::navigate(const BrowsingContext& browsingContext, const String& url, std::optional<Inspector::Protocol::BidiBrowsingContext::ReadinessState>&&, Ref<NavigateCallback>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR(NotImplemented);
```

Comment:
is there a point of even having this stub if it's not implemented?  i.e. would it be better to not have this here at all and instead wait for another PR that fully implements it?

Reply:
Yeah, I had a hacky implementation in my working copy but I dropped it since it will be rewritten soon. The reason that this stub exists is so that I could validate command generation for multiple domains.